### PR TITLE
Np 46529 UpsertNviCandidateHandler: Persist points for creator affiliation

### PR DESCRIPTION
**Pull request by monaullah**
Jira task: NP-46529, Refactor points for affiliation calculation, persist in db.

Part 1 -> [Merged](https://github.com/BIBSYSDEV/nva-nvi/pull/279)
Part 2 -> [Merged](https://github.com/BIBSYSDEV/nva-nvi/pull/283)
This is part 3: Persist points for affiliation in `UpsertNviCandidateHandler`

- Change in `EvalutatorService`: instead of sending affiliation points as part of `NviCreator`, change data type for `institutionPoints` from `Map<URI, BigDecimal>` to `List<InstitutionPoints>` (containing `CreatorAffiliationPoints`)
- Changed interface `UpsertCandidateRequest`: change data type for `institutionPoints` from `Map<URI, BigDecimal>` to `List<InstitutionPoints>` (containing `CreatorAffiliationPoints`)
- `Candidate` (BO): Persist creator affiliation points when persisting candidate in DB
- `CandidateDao`: Added `DbCreatorAffiliationPoints` in `DbInstitutionPoints`

This change will require a new point calculation for all candidates in all environments. I can use `ReEvaluateNviCandidatesHandler` for this purpose.

Also, tested in sandbox. See CANDIDATE#54a8ff9f-4a22-41d9-9c1d-f47031f1eef4 in DB.

(Sorry about the size, a lot of tests where affected)

### CODE REVIEW

File: event-handlers/src/main/java/no/sikt/nva/nvi/events/cristin/CristinMapper.java [hunk 1]
```diff
@@ -248,7 +248,8 @@ private static boolean hasInstitutionPoints(ScientificPerson scientificPerson) {
     }
 
     private static DbInstitutionPoints toDbInstitutionPoints(Entry<URI, BigDecimal> entry) {
-        return new DbInstitutionPoints(entry.getKey(), entry.getValue());
+        //TODO: Implement creatorAffiliationPoints
+        return new DbInstitutionPoints(entry.getKey(), entry.getValue(), List.of());
```
Comment (monaullah) [inline event-handlers/src/main/java/no/sikt/nva/nvi/events/cristin/CristinMapper.java:252]: This is the only change here

Comment (monaullah) [reply event-handlers/src/main/java/no/sikt/nva/nvi/events/cristin/CristinMapper.java:252]: @truhacevkir fyi

File: event-handlers/src/test/java/no/sikt/nva/nvi/events/evaluator/EvaluateNviCandidateHandlerTest.java [hunk 2]
```diff
@@ -165,7 +167,8 @@ void shouldCreateNewCandidateWithPointsOnlyForNviInstitutions() throws IOExcepti
         var messageBody = getMessageBody();
         var institutionPoints = ((NviCandidate) messageBody.candidate()).institutionPoints();
         assertEquals(1, institutionPoints.size());
-        assertTrue(institutionPoints.containsKey(CRISTIN_NVI_ORG_TOP_LEVEL_ID));
+        assertTrue(institutionPoints.stream().map(
+            InstitutionPoints::institutionId).anyMatch(uri -> uri.equals(CRISTIN_NVI_ORG_TOP_LEVEL_ID)));
```
Comment (brinxmat) [inline event-handlers/src/test/java/no/sikt/nva/nvi/events/evaluator/EvaluateNviCandidateHandlerTest.java:171]: Is it an object access pattern to filter by URI for InstitutionPoints? If so, probably create a method for this.

File: event-handlers/src/test/java/no/sikt/nva/nvi/events/evaluator/EvaluateNviCandidateHandlerTest.java [hunk 3]
```diff
@@ -425,6 +428,14 @@ void shouldReadDlqAndRetryAfterGivenTime() {
 
     }
 
+    private static BigDecimal getInstitutionPoints(List<InstitutionPoints> institutionPoints, URI institutionId) {
+        return institutionPoints.stream()
+                   .filter(institutionPoint -> institutionPoint.institutionId().equals(institutionId))
```
Comment (brinxmat) [inline event-handlers/src/test/java/no/sikt/nva/nvi/events/evaluator/EvaluateNviCandidateHandlerTest.java:433]: You're accessing the uri again here…

File: event-handlers/src/test/java/no/sikt/nva/nvi/events/evaluator/EvaluateNviCandidateWithCristinDataTest.java [hunk 4]
```diff
@@ -140,8 +146,18 @@ void shouldReturnSamePointsAsPointsCalculatedByCristinForAcademicChapterFrom2022
...
-        assertThat(institutionPoints.get(SINTEF_TOP_LEVEL_ORG_ID), is(equalTo(scaledBigDecimal(0.5000))));
+        assertThat(getInstitutionPoints(institutionPoints, NTNU_TOP_LEVEL_ORG_ID),
+                   is(equalTo(scaledBigDecimal(0.8660))));
+        assertThat(getInstitutionPoints(institutionPoints, SINTEF_TOP_LEVEL_ORG_ID),
+                   is(equalTo(scaledBigDecimal(0.5000))));
+    }
+
+    private static BigDecimal getInstitutionPoints(List<InstitutionPoints> institutionPoints, URI institutionId) {
+        return institutionPoints.stream()
+                   .filter(institutionPoint -> institutionPoint.institutionId().equals(institutionId))
```
Comment (brinxmat) [inline event-handlers/src/test/java/no/sikt/nva/nvi/events/evaluator/EvaluateNviCandidateWithCristinDataTest.java:157]: Also here

File: nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/model/Candidate.java [hunk 5]
```diff
@@ -381,7 +381,11 @@ private static boolean pointsAreUpdated(UpsertCandidateRequest request, Candidat
     }
 
     private static BigDecimal extractRequestPoints(UpsertCandidateRequest request, URI institutionId) {
-        return request.institutionPoints().get(institutionId);
+        return request.institutionPoints().stream()
+                   .filter(institutionPoints -> institutionPoints.institutionId().equals(institutionId))
```
Comment (brinxmat) [inline nvi-commons/src/main/java/no/sikt/nva/nvi/common/service/model/Candidate.java:385]: Here…

File: event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/EvaluatorService.java [hunk 6]
```diff
@@ -50,7 +46,9 @@ public CandidateEvaluatedMessage evaluateCandidacy(URI publicationBucketUri) {
             publication);
         if (!verifiedCreatorsWithNviInstitutions.isEmpty()) {
             var pointCalculation = pointService.calculatePoints(publication, verifiedCreatorsWithNviInstitutions);
-            var nviCandidate = constructNviCandidate(publication, pointCalculation, publicationId,
+            var nviCandidate = constructNviCandidate(publication,
+                                                     verifiedCreatorsWithNviInstitutions, pointCalculation,
```
[the remark this answers is not in the document]
Comment (monaullah) [reply event-handlers/src/main/java/no/sikt/nva/nvi/events/evaluator/EvaluatorService.java:50]:
```
if (!verifiedCreatorsWithNviInstitutions.isEmpty()) {
    var pointCalculation = pointService.calculatePoints(publication, verifiedCreatorsWithNviInstitutions);
    var nviCandidate = ...
    return constructMessage(nviCandidate);
} else {
    return constructMessage(new NonNviCandidate(publicationId));
}
```

You hit return if List<VerifiedNviCreator> is empty?

**Pull request merged**